chain/client/src/client_actions.rs

```diff
@@ -1543,22 +1543,20 @@ impl ClientActions {
                         unwrap_and_report!(self.client.chain.get_block_header(&sync_hash));
                     let prev_hash = *block_header.prev_hash();
                     let epoch_id = block_header.epoch_id().clone();
-                    let shards_to_sync: Vec<_> = self
+                    let shards_to_sync = get_shards_cares_about_this_or_next_epoch(
+                        me.as_ref(),
+                        true,
+                        &block_header,
+                        &self.client.shard_tracker,
+                        self.client.epoch_manager.as_ref(),
+                    );
+                    let shard_uids = self
                         .client
                         .epoch_manager
-                        .shard_ids(&epoch_id)
-                        .unwrap()
-                        .into_iter()
-                        .filter(|&shard_id| {
-                            cares_about_shard_this_or_next_epoch(
-                                me.as_ref(),
-                                &prev_hash,
-                                shard_id,
-                                true,
-                                &self.client.shard_tracker,
-                            )
-                        })
-                        .collect();
+                        .shard_ids_to_uids(shards_to_sync.as_slice(), &epoch_id)
+                        .unwrap();
+                    // Make sure mem-tries can be loaded only for shards we care about this or next epoch.
+                    self.client.runtime_adapter.retain_mem_tries(shard_uids.as_slice());
```

Comment:
nit: do we need to `as_slice()` call here?

Reply:
Yes, compiler enforces it.